src/System.IO.Compression.Brotli/src/System/IO/Compression/BrotliStream.cs

```diff
@@ -40,13 +40,27 @@ public BrotliStream(Stream stream, CompressionMode mode, bool leaveOpen)
             _mode = mode;
             _stream = stream;
             _leaveOpen = leaveOpen;
-            _buffer = new byte[DefaultInternalBufferSize];
+            _buffer = ArrayPool<byte>.Shared.Rent(DefaultInternalBufferSize);
+        }
+
+        private void ReturnBufferToPool()
+        {
+            Debug.Assert(_stream == null, "Stream must be disposed to return the buffer.");
+            byte[] buffer = _buffer;
+            if (buffer != null)
+            {
+                _buffer = null;
+                if (!AsyncOperationIsActive)
```

Comment:
Why do we need this check?

And if we need it, shouldn't we avoid setting `_buffer` to null unless `AsyncOperationIsActive` is false?
Otherwise, won't we get null refs here:
https://github.com/dotnet/corefx/blob/266ba84eae00fcafa114efe7bdb397354e69dc06/src/System.IO.Compression.Brotli/src/System/IO/Compression/dec/BrotliStream.Decompress.cs#L138-L141

Reply:
> Why do we need this check?

Such that with misuse, e.g.
```C#
using (var stream = new BrotliStream(…))
{
     stream.ReadAsync(buffer, 0, buffer.Length);
}
```
we don't put the ArrayPool-rented buffer back into the pool while it's still in use.

> shouldn't we avoid setting _buffer to null unless AsyncOperationIsActive is false?

I'd rather always null it out.  I think it's fine to get a null reference exception if you use a type after it's been Disposed; I'm sure there are plenty of cases where that happens in other types in the framework.  But that's local corruption to a particular instance.  What I want to avoid here is global process corruption that's difficult to track down, spooky action at a distance with the same buffer in use in multiple places.  This is a similar pattern to what we do in DeflateStream.

Reply:
> we don't put the ArrayPool-rented buffer back into the pool while it's still in use.

And is it fine if that means we end up not returning the buffer at all, in that case? When will that resource get collected?

Also, out of curiosity, what error state would the program be in if we were to return the buffer back to the pool while it's being used, anyway?

Reply:
> And is it fine if that means we end up not returning the buffer at all, in that case?

Yes, you've got a bug in your program; it's ok to not return a buffer to the pool.

>  When will that resource get collected?

It'll just get collected by the GC like any other array.

> what error state would the program be in if we were to return the buffer back to the pool while it's being used, anyway?

Multiple threads all using the buffer concurrently, reading and writing from it, corrupting each other's data, etc.